proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,461 @@
+
+Of course, REPLs are another great way to get started. But they are outside the scope of this document.
+
+## What is a Simple C# Program
+
+A Simple C# Program is this:
+
+* A C# file
+* A set of global `using`s inferred by the build environment
+* A tool that can run that file / produce a working program
+* A way to incorporate libraries and frameworks/sdks into that file and run it
+* A way to "graduate" to using a project file
+
+A Simple C# Program is **not** a replacement for project-based C# code or .NET Interactive.
+
+## Simple C# is just C# code
+
+Simple C# Programs are not a dialect of the C# language, like the interactive dialect, CSX. Any code in a Simple C# Program can be built and run in a project-based program.
+
+## User experience walkthrough
+
+Given the following code in a singular `Hello.cs` in a directory:
+
+```csharp
+WriteLine("Hello World");
+```
+
+You can just run it:
+
+```bash
+$ dotnet run
+Hello World
+```
+
+Or with a `#!` directive,
+
+```csharp
+#!/path/to/dotnet
+WriteLine("Hello World");
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+Hello World
+```
+
+When running via `dotnet run` or the `#!` directive, artifacts may be created on disk in the temp folder. They could also be generated in memory.
+
+The `using`s that are normally in a C# project template are implicitly available at design-time. This is accomplished with [Global Using Directives](https://github.com/dotnet/csharplang/blob/main/proposals/GlobalUsingDirective.md) that the build tooling will have available.
+
+The `#!` directive is covered in [Hashbang (#!) support](https://github.com/dotnet/csharplang/issues/3507).
+
+### Building instead of running
+
+You can also build it with `dotnet build`:
+
+```bash
+$ dotnet build
+$ ls
+Hello.exe
+$ ./Hello.exe
+Hello World
+```
+
+When built, it creates a single executable that it drops in the current working directory.
+
+### Referencing a package
+
+You can also pull in a package (syntax perhaps subject to change):
+
+```csharp
+#r "nuget: Newtonsoft.Json"
+
+record Person(string Name);
+
+var json = "{\"name\": \"dotnet bot\"}";
+var person = JsonConvert.DeserializeObject<Person>(json);
+
+Console.WriteLine(person.Name);
+```
+
+And it'll run after downloading the package:
+
+```bash
+$ dotnet run
+dotnet bot
+```
+
+Or via a `#!` directive,
+
+```csharp
+#!/path/to/dotnet
+#r "nuget: Newtonsoft.Json"
+
+record Person(string Name);
+
+var json = "{\"name\": \"dotnet bot\"}";
+var person = JsonConvert.DeserializeObject<Person>(json);
+
+Console.WriteLine(person.Name);
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+dotnet bot
+```
+
+The `#r "nuget:..."` syntax is already supported by C# today in .NET Interactive. It is a shared implementation between C# and F#. The syntax and mechanism could be re-used here.
```

Comment:
I think the proposal if for C#-proper to tolerate but ignore such directives. Those directives would be picked up by the SDK to use in synthesized project file.
Let's confirm in LDM on Wednesday